src/Symfony/Component/Messenger/Middleware/HandleRecordedMessageMiddleware.php

```diff
@@ -0,0 +1,71 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Messenger\Middleware;
+
+use Symfony\Component\Messenger\Exception\MessageHandlingException;
+use Symfony\Component\Messenger\MessageBusInterface;
+use Symfony\Component\Messenger\RecordedMessageCollectionInterface;
+
+/**
+ * A middleware that takes all recorded messages and dispatch them to the bus.
+ *
+ * @author Tobias Nyholm <tobias.nyholm@gmail.com>
+ * @author Matthias Noback <matthiasnoback@gmail.com>
+ */
+class HandleRecordedMessageMiddleware implements MiddlewareInterface
+{
+    private $messageRecorder;
+    private $messageBus;
+
+    public function __construct(MessageBusInterface $messageBus, RecordedMessageCollectionInterface $messageRecorder)
+    {
+        $this->messageRecorder = $messageRecorder;
+        $this->messageBus = $messageBus;
+    }
+
+    public function handle($message, callable $next)
+    {
+        // Make sure the recorder is empty before we begin
+        $this->messageRecorder->resetRecordedMessages();
```

Comment:
Let's imagine the following example:

1. `Controller` `dispatch`es a `FirstMessage` message.
2. `FirstMessageHandler` gets called and it does the following:
    ```php
    $recorder->record(new SecondMessage);
    $bus->dispatch(new SaveToSecondaryDatabase);
    $recorder->record(new ThirdMessage);
    ```

The `dispatch` will call the middleware and the just recorded message will be erased, right?

Reply:
This is correct IF the current bus and `$bus` is using the same `MessageRecorder`. 

I think I should remove this line. Are you okey with that @ogizanagi?  It was added by [your suggestion](https://github.com/symfony/symfony/pull/27844#discussion_r200204703).

Reply:
It might be fine removing it, but the case described above is a more complex topic then.
For instance, if there is any issue with `FirstMessageHandler` after `SaveToSecondaryDatabase`, `SecondMessage` 
 (and possibly those recorded by `SaveToSecondaryDatabase`) would still have been executed while I would expect it does only if `FirstMessageHandler` is fully successful.
Actually, to me recorded messages should be treated if and only if the highest `dispatch` call is successful. Possibly by using an internal flag inside the middleware? (Thus, we could even keep the reset for safety, only when the flag is off).

Reply:
Perhaps the `LockingMiddleware` of Tactician could be a source of inspiration: 
- http://tactician.thephpleague.com/plugins/locking-middleware/
- https://github.com/thephpleague/tactician/blob/master/src/Plugins/LockingMiddleware.php

Reply:
Which is equivalent to [the `FinishesHandlingMessageBeforeHandlingNext` middleware](https://github.com/SimpleBus/SimpleBus/blob/master/Component/MessageBus/src/Bus/Middleware/FinishesHandlingMessageBeforeHandlingNext.php) of SimpleBus. (I prefer SimpleBus' naming in here, it's more explicit)